tools/codegen/api/cpp.py

```diff
@@ -191,9 +195,89 @@ def argument(a: Union[Argument, TensorOptionsArguments, ThisArgument]) -> CppArg
+
+    def cpp_arguments(self) -> Sequence[CppArgument]:
+        return list(map(argument, self.arguments))
+
+    # Return arguments as a comma separated list, i.e. like they would be in a C++
+    # function signature. Include default values for arguments.
+    def cpp_arguments_str(self) -> str:
+        return ', '.join(map(str, self.cpp_arguments()))
+
+    # Return a copy of the CppSignature with the `this` argument removed
+    def exclude_this(self) -> 'CppSignature':
+        return CppSignature(
+            arguments=[a for a in self.arguments if not isinstance(a, ThisArgument)]
+        )
+
+    # Return a copy of the CppSignature with all default values removed from the signature
+    def no_defaults(self) -> 'CppSignature':
+        arguments: List[Union[Argument, TensorOptionsArguments, ThisArgument]] = []
+        for argument in self.arguments:
+            if isinstance(argument, Argument) or isinstance(argument, ThisArgument):
+                arguments.append(dataclasses.replace(
+                    argument,
+                    default=None))
+            elif isinstance(argument, TensorOptionsArguments):
+                arguments.append(dataclasses.replace(
+                    argument,
+                    dtype=dataclasses.replace(argument.dtype, default=None),
+                    layout=dataclasses.replace(argument.layout, default=None),
+                    device=dataclasses.replace(argument.device, default=None),
+                    pin_memory=dataclasses.replace(argument.pin_memory, default=None),
+                ))
+            else:
+                raise "Unhandled case"
+        return CppSignature(arguments=arguments)
+
+    # Return a string with a comma separated list of expressions that could be used
+    # to call this operator. This can be used to generate code that wraps operators
+    # and calls back into them. The process_tensoroptions argument determines how
+    # tensor options should be treated. They can be
+    # - PASS_THROUGH: Don't do anything, just handle them as regular arguments
+    # - SCATTER: Expect a `TensorOptions options` in the scope and scatter it into `options.dtype, ...`
+    # - GATHER: Expect `dtype, ...` in the scope and gather them into a TensorOptions for calling
+    def exprs_str(self,
+                  process_tensoroptions: dispatcher.ProcessTensoroptions = dispatcher.ProcessTensoroptions.PASS_THROUGH
+                  ) -> str:
+        args = self.cpp_arguments()
+        exprs = dispatcher.cpparguments_exprs(args, process_tensoroptions=process_tensoroptions)
+        return ', '.join(map(lambda a: a.expr, exprs))
+
+    def types_str(self) -> str:
+        args = self.cpp_arguments()
+        exprs = dispatcher.cpparguments_exprs(args, process_tensoroptions=dispatcher.ProcessTensoroptions.PASS_THROUGH)
+        return ', '.join(map(lambda a: a.type, exprs))
+
+@dataclass(frozen=True)
+class CppBinding:
```

Comment:
nit: naming. Binding to me implies "we are binding some arguments into names". But this class doesn't do any binding at all, it's just a collection of signatures. Call it CppSignatures or CppSignatureGroup.

Reply:
I renamed it to `CppSignatureGroup`.